File: Editorial/98_Validate_Binary_Search_Tree/C/1_Recursive_Traversal_with_Valid_Range.c
bool validate(struct TreeNode* root, struct TreeNode* low,
              struct TreeNode* high) {
    // Empty trees are valid BSTs.
    if (root == NULL) {
        return true;
    }
    // The current node's value must be between low and high.
    if ((low != NULL && root->val <= low->val) ||
        (high != NULL && root->val >= high->val)) {
        return false;
    }
    // The left and right subtree must also be valid.
    return validate(root->right, root, high) && validate(root->left, low, root);
}
bool isValidBST(struct TreeNode* root) { return validate(root, NULL, NULL); }

File: Editorial/0000-0100/0098_Validate_Binary_Search_Tree/JavaScript/1_Recursive_Traversal_with_Valid_Range.js
var isValidBST = function (root, low = null, high = null) {
    // Empty trees are valid BSTs.
    if (!root) return true;
    // The current node's value must be between low and high.
    if ((low != null && root.val <= low) || (high != null && root.val >= high))
        return false;
    // The left and right subtree must also be valid.
    return (
        isValidBST(root.right, root.val, high) &&
        isValidBST(root.left, low, root.val)
    );
};
